feat(upgrade): provide test helpers for wiring up injectors

petebacondarwin opened the pull request:
**Sits on top of #29466**

This PR provides two test helpers for the upgrade/static library, which wire up the Angular and AngularJS injectors without the need for a full bootstrap of a hybrid app.

* `createAngularJSTestingModule()` should be used in AngularJS tests
* `createAngularTestingModule()` should be used in Angular tests

Check out the spec files for example usage.

mary-poppins commented:
The angular.io preview for bfe42050247959ce6d254c82cbff83eef1b12bbf is available [here][1].

[1]: https://pr16848-bfe42050247959ce6d254c82cbff83eef1b12bbf.ngbuilds.io/

mary-poppins commented:
The angular.io preview for c73b3926cbe9318a4eb56fae092d69a67e916cd0 is available [here][1].

[1]: https://pr16848-c73b3926cbe9318a4eb56fae092d69a67e916cd0.ngbuilds.io/

mhevery commented:
Looking good! I think we need some more documentation on how to use this. Can this be part of the API docs?

petebacondarwin commented:
I'll add docs but wanted feedback on whether this is the right API first.

juliemr commented on packages/upgrade/test/testing/create_angular_test_module_spec.ts:
```diff
@@ -0,0 +1,27 @@
+ */
+
+import {TestBed} from '@angular/core/testing';
+import * as angular from '@angular/upgrade/src/common/angular1';
```
Is this line required?

petebacondarwin replied:
You're right. It is not needed.

petebacondarwin replied:
I guess that I should also import the AngularJS app module from `mocks` and use its name in line 19 below, rather than hard coding it.

gkalpak commented on packages/upgrade/src/testing/create_angular_test_module.ts:
```diff
@@ -0,0 +1,31 @@
+  constructor(i: Injector) { injector = i; }
+}
+
+export function createAngularTestingModule(angularJSModules: string[], strictDI = true) {
```
Why `string[]`? `angularJSModules` can contain functions or arrays (which is not common in production code, but it is in tests).

petebacondarwin replied:
I am going to leave it as just strings for now. We can always update it later if it is problem.

gkalpak commented on packages/upgrade/src/testing/create_angularjs_test_module.ts:
```diff
@@ -0,0 +1,30 @@
+import {$INJECTOR, INJECTOR_KEY} from '../common/constants';
+
+export function createAngularJSTestingModule(AngularModules: any[]) {
+  return angular.module('$$angularJSTestingModule', [])
```
Would it be possible/useful to use both `createXyzTetingModule` helpers in the same test (in which case we need different AngularJS module names)?

petebacondarwin replied:
I don't think you could/should use both helpers in the same test. One is designed for running tests from inside the AngularJS test infrastructure and the other from the Angular infrastructure.

gkalpak commented on packages/upgrade/test/testing/create_angularjs_test_module_spec.ts:
```diff
@@ -0,0 +1,26 @@
+import {createAngularJSTestingModule} from '@angular/upgrade/src/testing/create_angularjs_test_module';
+import {AppModule, Inventory} from './mocks';
+
+const ngMock = (window as any).angular.mock;
```
Why not ue the imported `angular` or even add `mock` as an optional property and import just that?

petebacondarwin replied:
I changed this to use `getAngularJSGlobal()`

gkalpak commented:
As a reminder, the helpers need to be exported and `upgrade/testing` should be added to a bunch of lists for it to appear in the docs etc.

petebacondarwin commented on packages/upgrade/src/testing/create_angular_test_module.ts:
```diff
@@ -0,0 +1,31 @@
+  constructor(i: Injector) { injector = i; }
+}
+
+export function createAngularTestingModule(angularJSModules: string[], strictDI = true) {
```
Fair point

petebacondarwin replied:
Actually, this is slightly different since we are not calling through to the `ngMock.module()` function but are actually calling the `angular.module()` function, which does indeed only expect an array of strings for the dependencies.

gkalpak replied:
AFAICT, it does accept functions or arrays :confused:
(And they are useful for tests :grin:)

petebacondarwin replied:
We are talking about this parameter: https://github.com/angular/angular.js/blob/master/src/loader.js#L75

Are you saying that one can do:

```javascript
angular.module('app', [function($rootScope) { ... }]);
```

gkalpak replied:
Actually, you can, but I know realize that un-annotated functions will not work correctly when `strict-di` is, unless passed thought `angular.mock.module`.

I guess there is a reason for using `angular.module` directly, right?

petebacondarwin commented on packages/upgrade/test/testing/create_angularjs_test_module_spec.ts:
```diff
@@ -0,0 +1,26 @@
+import {createAngularJSTestingModule} from '@angular/upgrade/src/testing/create_angularjs_test_module';
+import {AppModule, Inventory} from './mocks';
+
+const ngMock = (window as any).angular.mock;
```
Sadly we can't. I did try. The `angular` that we import is not actually the global. It is a local export from our `common/angular1.ts` file, which manually delegates to the global (once it has been "set" via `setAngularJSLib()`).

I couldn't come up with a simpler way to achieve this. Any ideas?

gkalpak replied:
We could expose `mock` (if present on the `angular` object) from `angular1.ts`. But given this is just for our tests and not needed in `ngUpgrade` itself, getting it from `window.angular` is fine imo.

petebacondarwin replied:
That was also my view too

juliemr commented:
Ping?

petebacondarwin commented:
Sorry I have been distracted. I will tidy it up this week.

petebacondarwin commented:
I have made the suggested changes, where appropriate, added documentation and included this new module in as many configuration files as I could find.

PTAL: @mhevery @juliemr @gkalpak

petebacondarwin commented:
And of course this PR falls foul of the change from CircleCI v1 to v2.

petebacondarwin commented:
Just checking to see if closing and reopening helps with CircleCI

mary-poppins commented:
You can preview 52735ee at https://pr16848-52735ee.ngbuilds.io/.

mary-poppins commented:
You can preview 0743841 at https://pr16848-0743841.ngbuilds.io/.

gkalpak commented on packages/upgrade/testing/src/create_angular_test_module.ts:
```diff
@@ -0,0 +1,67 @@
+/**
```
Nit: We could keep the names consistent with the exported function names: `_test_module` --> `_testing_module` :grin:

gkalpak reviewed (commented):
Just a few nits. Looks great so far (except that I have no idea about all the places we need to list a new package :grin:)
I assume more tests will be added when this is finalized.

One thing I am a little concerned with is the namespacing of the module. Typically `@angular/foo/testing` is for testing `@angular/foo/bar`. So, I would expect the testing package for `@angular/upgrade/static` to be `@angular/upgrade/static/testing` (while `@angular/upgrade/testing` sounds more about testing `@angular/upgrade`).
I hope that eventually, `@angular/upgrade` will get dropped and `@angular/upgrade/static` will be moved to `@angular/upgrade` (together with its testing module).

petebacondarwin commented on aio/tools/transforms/angular-api-package/index.js:
```diff
@@ -57,6 +57,7 @@ module.exports = new Package('angular-api', [basePackage, typeScriptPackage])
       'router/upgrade/index.ts',
       'upgrade/index.ts',
       'upgrade/static/index.ts',
+      'upgrade/testing/index.ts'
```
But is it? All the code that we reference is in the `common` folder, so I believe that this `testing` module could be used for either dynamic or static. What do you think?

gkalpak replied:
Good point. From a quick look it seems you are right (in which case the name is fine) :smiley:

petebacondarwin commented on packages/upgrade/testing/src/create_angular_test_module.ts:
````diff
@@ -0,0 +1,67 @@
+ * In the following code snippet, we are configuring the TestBed with two imports.
+ * The `AppModule` is the Angular part of our hybrid application and the `appModule` is the AngularJS part.
+ *
+ * ```
````
Not sure but I will add the language for good measure

petebacondarwin commented on packages/upgrade/testing/src/create_angularjs_test_module.ts:
````diff
@@ -0,0 +1,61 @@
+ * In the following code snippet, `shoppingCart` is an AngularJS service that depends upon an Angular service.
+ *
+ * ```
+ * it('should ...', inject(['shoppingCart', function(shoppingCart) {
````
fair enough. I was trying to make it fail proof :-)

petebacondarwin commented on packages/upgrade/testing/src/create_angularjs_test_module.ts:
````diff
@@ -0,0 +1,61 @@
+ * ```
+ *
+ * @param angularJSModules a collection of the names of AngularJS modules to include in the configuration
+ * @param strictDI whether the AngularJS injector should have `strictDI` enabled. Defaults to `true`.
````
:scream:

kieranblight commented:
Hi Pete, I am just wondering what the rough timeline for this to be release will be?

mary-poppins commented:
You can preview 0260634 at https://pr16848-0260634.ngbuilds.io/.

petebacondarwin commented:
@kieranblight - I have been having difficulties getting the new package into the system. I hope to finish this and get it in by the end of this week.

kieranblight commented:
@petebacondarwin thank you, look forward to it :)!

juliemr commented on packages/upgrade/testing/src/create_angular_test_module.ts:
```diff
@@ -0,0 +1,67 @@
+export function createAngularTestingModule(angularJSModules: string[], strictDI = true) {
+  angular.module('$$angularJSTestingModule', angularJSModules)
+      .factory(INJECTOR_KEY, () => injector);
+  $injector = angular.injector(['$$angularJSTestingModule'], strictDI);
```
This should probably include `ng` as in:

`$injector = angular.injector(['ng', '$$angularJSTestingModule'], strictDI);`

So that createAngularTestingModule works without explicitly listing ng?

gkalpak replied:
For context:
This is automatically done when [bootstrapping an app][1] or by `ngMock` in tests. So we must "manually" do it here.

[1]: https://github.com/angular/angular.js/blob/87a586eb9a23cfd0d0bb681cc778b4b8e5c8451d/src/Angular.js#L1815
[2]: https://github.com/angular/angular.js/blob/87a586eb9a23cfd0d0bb681cc778b4b8e5c8451d/src/ngMock/angular-mocks.js#L3138

petebacondarwin commented on packages/upgrade/testing/src/create_angularjs_test_module.ts:
```diff
@@ -0,0 +1,61 @@
+ * @param angularJSModules a collection of the names of AngularJS modules to include in the configuration
+ * @param strictDI whether the AngularJS injector should have `strictDI` enabled. Defaults to `true`.
+ */
+export function createAngularJSTestingModule(AngularModules: any[]) {
```
The `TestBed.configureTestingModule()` function expects `imports` to be `any[]`

petebacondarwin commented on packages/upgrade/testing/src/create_angularjs_test_module.ts:
```diff
@@ -0,0 +1,61 @@
+          INJECTOR_KEY,
+          [
+            $INJECTOR,
+            ($injector: any) => {
```
fixed

mary-poppins commented:
You can preview b0a0213 at https://pr16848-b0a0213.ngbuilds.io/.

mary-poppins commented:
You can preview 3808f4a at https://pr16848-3808f4a.ngbuilds.io/.

mary-poppins commented:
You can preview 0ca84ba at https://pr16848-0ca84ba.ngbuilds.io/.

mary-poppins commented:
You can preview 733d0aa at https://pr16848-733d0aa.ngbuilds.io/.

mary-poppins commented:
You can preview 98094af at https://pr16848-98094af.ngbuilds.io/.

mary-poppins commented:
You can preview 37b041c at https://pr16848-37b041c.ngbuilds.io/.

mary-poppins commented:
You can preview 391d3b6 at https://pr16848-391d3b6.ngbuilds.io/.

mary-poppins commented:
You can preview 0e7ae4a at https://pr16848-0e7ae4a.ngbuilds.io/.

mary-poppins commented:
You can preview 78b3b07 at https://pr16848-78b3b07.ngbuilds.io/.

mary-poppins commented:
You can preview dd3997d at https://pr16848-dd3997d.ngbuilds.io/.

petebacondarwin commented:
Agh! Something changed in the way that paths are configured in tsconfig.json, which means that the builds now break inside the `aio_e2e` tests fail... E.g https://travis-ci.org/angular/angular/jobs/271671271

Any ideas @jasonaden ?

mary-poppins commented:
You can preview bbb982d at https://pr16848-bbb982d.ngbuilds.io/.

calebeno commented:
@mary-poppins Is there an eta for this being merged in?

petebacondarwin commented:
Il look at this again this week

mary-poppins commented:
You can preview 12c97dd at https://pr16848-12c97dd.ngbuilds.io/.

mary-poppins commented:
You can preview fc00d70 at https://pr16848-fc00d70.ngbuilds.io/.

mary-poppins commented:
You can preview c9d70db at https://pr16848-c9d70db.ngbuilds.io/.

mary-poppins commented:
You can preview 3010b5d at https://pr16848-3010b5d.ngbuilds.io/.

mary-poppins commented:
You can preview 44a5fda at https://pr16848-44a5fda.ngbuilds.io/.

mary-poppins commented:
You can preview b8eedda at https://pr16848-b8eedda.ngbuilds.io/.

petebacondarwin commented:
I think this is ready to go now. Just needs some approval...

gkalpak commented on packages/upgrade/static/testing/tsconfig-build.json:
```diff
@@ -0,0 +1,26 @@
+      "@angular/common": ["../../../../dist/packages/common"],
+      "@angular/compiler": ["../../../../dist/packages/compiler"],
+      "@angular/platform-browser": ["../../../../dist/packages/platform-browser"],
+      "@angular/platform-browser-dynamic": ["../../../../dist/packages/platform-browser-dynamic"]
```
Do we need all these paths?

petebacondarwin replied:
Nope

gkalpak commented on packages/upgrade/static/testing/tsconfig-build.json:
```diff
@@ -0,0 +1,26 @@
+  },
+  "files": [
+    "public_api.ts",
+    "../../../../node_modules/zone.js/dist/zone.js.d.ts"
```
Is this needed?

petebacondarwin replied:
still need to check this...

petebacondarwin replied:
Removed.

gkalpak commented on packages/upgrade/static/testing/tsconfig-build.json:
```diff
@@ -0,0 +1,26 @@
+  ],
+  "angularCompilerOptions": {
+    "annotateForClosureCompiler": true,
+    "strictMetadataEmit": true,
```
Why different than [core/testing/tsconfig-build.json][1]?

[1]: https://github.com/petebacondarwin/angular/blob/b8eedda6dd41aec47ec3aa17d5aaee70acbc676f/packages/core/testing/tsconfig-build.json#L21-L22

petebacondarwin replied:
It should now look pretty much the same.

calebeno commented:
@petebacondarwin Sorry to comment again, but this would be an incredibly useful tool to have for testing purposes.  Would you have time to address the review comments?

petebacondarwin commented:
Yes, tomorrow :-) I promise.

calebeno commented on packages/upgrade/static/testing/src/create_angular_testing_module.ts:
```diff
@@ -0,0 +1,74 @@
+  angular.module('$$angularJSTestingModule', angularJSModules)
+      .factory(INJECTOR_KEY, () => injector);
+  $injector = angular.injector(['ng', '$$angularJSTestingModule'], strictDI);
+  return AngularTestingModule;
```
We've been copying some of this code into our test utilities so we can use it before this gets merged in.  I can confirm that this is definitely an issue.  Specifically, any test files using `createAngularJSTestingModule` after the one that uses `createAngularTestingModule`  fail all specs with `No provider for` errors.

petebacondarwin replied:
@gkalpak I'll have a play around but I never intended for this helper to be called twice in a single test. If you want multiple NgModules to be used in a test then I think they should be imported into the Angular NgModule created by `TestBed.configureTestingModule`; if you want multiple AngularJS modules to be used in a test then they should all be listed in the first parameter (`angularJSModules`) of the call to `createAngularTestingModule`.

Note that the AngularJS injector is a global (exported by this ES6 module). So if we wanted to create multiple NgModules, then we would need to decide whether they share the same AngularJS injector or not. This seems a complicated testing setup, when I believe that using `imports` on the single testing module would be more appropriate...

petebacondarwin replied:
@calebeno - my intention was that these two utilities (`createAngularJSTestingModule` and `createAngularTestingModule`) would not be used in the same test run. The first is to allow AngularJS services to be used when testing Angular services; the second is to allow Angular services to be used when testing AngularJS services.

I don't understand how they can be conflicting unless you are using them together - i.e. calling both helpers and passing one to the other. For example:

```
createAngularTestingModule([appModule.name], module(createAngularJSTestingModule([AppModule]))
```

Can you provide more info on what you are doing, so I can investigate?

petebacondarwin replied:
The global `$injector` and `injector` are a bit of problem. I am not quite sure how we could get around that:

* the `$injector` needs to be available to the `NgModule` but we don't get to instantiate that class so not sure how we could provide it
* the AngularJS module has a factory that needs to have access to the Angular `injector` but this need a reference to a variable for `injector` from inside the `createAngularTestingModule` function.
* as long as you only use one call to either of these methods per test then there should be no leakage of data between tests, since the new call will effectively rewire things and cause the injectors to be reassigned to their global variables.
* there is potential for memory leakage, since the two injectors' variables are never unbound; but at least they do get overwritten reliably when a new testing module is created - which would mean that we only ever have one injector stuck in memory. We could provide a clear up method for this, but it doesn't appear to me that `TestBed` itself is doing much different...

gkalpak replied:
I am fine with the "once per test" restriction (for now 😛) as long as it is clearly documented.

But why do the injector vars have to be globals? Couldn't you create the modules inside the `createAngularTestingModule()` function? Would it break with aot (I wonder if ngUpgrade works with aot testing anyway 😕).

Alternatively, could you have an array of `injector`/`$injector` vars (if you really wanted to support multiple calls per test)?

petebacondarwin commented on packages/upgrade/static/testing/src/create_angular_testing_module.ts:
```diff
@@ -0,0 +1,74 @@
+ *
+ * import {appModule, AppModule, Inventory} from './app';
+ *
+ * TestBed.configureTestingModule({
```
That sounds correct.

petebacondarwin commented on packages/upgrade/static/testing/src/create_angular_testing_module.ts:
```diff
@@ -0,0 +1,74 @@
+ *
+ * @experimental
+ */
+export function createAngularTestingModule(angularJSModules: string[], strictDI = true) {
```
Hmm, good question. I was thinking it would be good practice. But I guess in tests you might not care - especially if you are using an automatic annotation tool?

petebacondarwin commented on packages/upgrade/static/testing/src/create_angularjs_testing_module.ts:
````diff
@@ -0,0 +1,63 @@
+ * import {appModule, AppModule} from './app';
+ *
+ * beforeEach(module(createAngularJSTestingModule([AppModule]), appModule.name));
+ * ```
````
I'll move to real files

petebacondarwin commented on packages/upgrade/static/testing/test/mocks.ts:
```diff
@@ -0,0 +1,75 @@
+
+@Injectable()
+export class Inventory {
+  constructor(@Inject('serverRequest') public serverRequest: any) { debugger; }
```
😱

petebacondarwin commented on tools/public_api_guard/upgrade/static/testing.d.ts:
```diff
@@ -0,0 +1,5 @@
+export declare function createAngularJSTestingModule(AngularModules: any[]): string;
+
+/** @experimental */
+export declare function createAngularTestingModule(angularJSModules: string[], strictDI?: boolean): typeof AngularTestingModule;
```
We don't really care about this return type. How about I change the API to return `Type<any>`?

petebacondarwin replied:
Need to reproduce public_api golden file

petebacondarwin commented on packages/upgrade/static/testing/rollup.config.js:
```diff
@@ -0,0 +1,28 @@
+  format: 'umd',
+  exports: 'named',
+  moduleName: 'ng.upgrade.static.testing',
+  plugins: [resolve()],
```
I guess the goalposts keep changing and I am not quick enough to keep up :-)

mary-poppins commented:
You can preview be9c789 at https://pr16848-be9c789.ngbuilds.io/.

mary-poppins commented:
You can preview 8b344a1 at https://pr16848-8b344a1.ngbuilds.io/.

calebeno commented:
@petebacondarwin Hey, sorry it took me a few days to get back with you.  I ended up adding a flag to our karma config which allows us to run either the ngjs unit tests or the ng ones.  Running them separately seems to solve the issue with using both `createAngularJSTestingModule` and `createAngularTestingModule`.  It's really tough to distill down an example as the application I'm working on has 200+ components/services.  Eventually, we'll reach a point where all of the ngjs stuff is converted over to ng and then we can dump all of this code anyway, so it's a pain point we can live with for the time being.  Thank you so much for providing this utility.  It's a huge help!  We'll be glad to pull in your actual code (instead of our copy/pasted version) once you get it merged and a new release comes out.

petebacondarwin reviewed (commented):
Updated after @gkalpak's review.

mary-poppins commented:
You can preview 4292132 at https://pr16848-4292132.ngbuilds.io/.

mary-poppins commented:
You can preview 68ff3e8 at https://pr16848-68ff3e8.ngbuilds.io/.

mary-poppins commented:
You can preview 74a0888 at https://pr16848-74a0888.ngbuilds.io/.

mary-poppins commented:
You can preview f06ed88 at https://pr16848-f06ed88.ngbuilds.io/.

petebacondarwin commented:
Hurray! I believe the failures on Travis are only due to the new Chrome release messing up our e2e test runs.

mary-poppins commented:
You can preview 5222693 at https://pr16848-5222693.ngbuilds.io/.

mary-poppins commented:
You can preview 98c73c4 at https://pr16848-98c73c4.ngbuilds.io/.

mary-poppins commented:
You can preview 74ea5a6 at https://pr16848-74ea5a6.ngbuilds.io/.

mary-poppins commented:
You can preview 9b3a034 at https://pr16848-9b3a034.ngbuilds.io/.

calebeno commented:
@petebacondarwin Been a while since an update.  Any news?

mary-poppins commented:
You can preview 131be9e at https://pr16848-131be9e.ngbuilds.io/.

petebacondarwin commented:
Rebased

mary-poppins commented:
You can preview ff417be at https://pr16848-ff417be.ngbuilds.io/.

mary-poppins commented:
You can preview 81a07a1 at https://pr16848-81a07a1.ngbuilds.io/.

calebeno commented:
@petebacondarwin Hello.  Just checking in to see what the eta on this might be.

petebacondarwin commented:
@calebeno - I will rebase again (if I can work out how to do the public API stuff now that the gulp task has gone - cc @alexeagle) and then chase the Googlers to review this.

mary-poppins commented:
You can preview ffbf0a9 at https://pr16848-ffbf0a9.ngbuilds.io/.

petebacondarwin commented:
Rebased again. I don't think we can land this during the RC period of the 6.0.0 release though.

pburgmer commented:
Hi,

I now have the problem that this pull request would solve ...
Would be great to see this in Angular 6.1

pburgmer commented:
Hi,

I don't know if this is the right place to report errors. If not, I'm sorry.

I tried createAngularJSTestingModule with Angular 6.0.0. I copied the function and replaced the constants with their string value. I had to type $injector parameter as any but that shouldn't matter I guess.

I used the function as described within a beforeEach to load an Angular Module.

Now I get this error:
`debug.js:21 TypeError: token.toString is not a function
    at stringify (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:283)
    at identifierName (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:793)
    at JitEmitterVisitor.push../node_modules/@angular/compiler/fesm5/compiler.js.JitEmitterVisitor.visitExternalExpr (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:22445)
    at ExternalExpr.push../node_modules/@angular/compiler/fesm5/compiler.js.ExternalExpr.visitExpression (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:8418)
    at :9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:12386
    at JitEmitterVisitor.push../node_modules/@angular/compiler/fesm5/compiler.js.AbstractEmitterVisitor.visitAllObjects (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:12418)
    at JitEmitterVisitor.push../node_modules/@angular/compiler/fesm5/compiler.js.AbstractEmitterVisitor.visitLiteralMapExpr (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:12384)
    at LiteralMapExpr.push../node_modules/@angular/compiler/fesm5/compiler.js.LiteralMapExpr.visitExpression (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:8636)
    at :9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:12399
    at JitEmitterVisitor.push../node_modules/@angular/compiler/fesm5/compiler.js.AbstractEmitterVisitor.visitAllObjects (:9876/_karma_webpack_/webpack:/node_modules/@angular/compiler/fesm5/compiler.js:12418)`

Token is an object and looks like this
`{
    $$angularJSTestingModule: {_invokeQueue: Array(1), _configBlocks: Array(0), _runBlocks: Array(0), info: ƒ, requires: Array(0), …}
    ng: {_invokeQueue: Array(0), _configBlocks: Array(1), _runBlocks: Array(0), info: ƒ, requires: Array(1), …}
    ngLocale: {_invokeQueue: Array(0), _configBlocks: Array(1), _runBlocks: Array(0), info: ƒ, requires: Array(0), …}
    ngMock: {_invokeQueue: Array(1), _configBlocks: Array(1), _runBlocks: Array(0), info: ƒ, requires: Array(1), …}
    pascalprecht.translate: {_invokeQueue: Array(12), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(1), …}
    reportit.core.api: {_invokeQueue: Array(3), _configBlocks: Array(1), _runBlocks: Array(0), info: ƒ, requires: Array(0), …}
    reportit.core.applicationProperties: {_invokeQueue: Array(2), _configBlocks: Array(0), _runBlocks: Array(0), info: ƒ, requires: Array(1), …}
    reportit.core.filter: {_invokeQueue: Array(6), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(6), …}
    reportit.core.metadata: {_invokeQueue: Array(1), _configBlocks: Array(0), _runBlocks: Array(0), info: ƒ, requires: Array(1), …}
    reportit.core.statistics: {_invokeQueue: Array(1), _configBlocks: Array(0), _runBlocks: Array(0), info: ƒ, requires: Array(1), …}
    reportit.core.user: {_invokeQueue: Array(2), _configBlocks: Array(0), _runBlocks: Array(0), info: ƒ, requires: Array(2), …}
    reportit.core.utils: {_invokeQueue: Array(9), _configBlocks: Array(0), _runBlocks: Array(2), info: ƒ, requires: Array(1), …}
    ui.router: {_invokeQueue: Array(2), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(3), …}
    ui.router.angular1: {_invokeQueue: Array(0), _configBlocks: Array(0), _runBlocks: Array(0), info: ƒ, requires: Array(0), …}
    ui.router.init: {_invokeQueue: Array(1), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(0), …}
    ui.router.router: {_invokeQueue: Array(1), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(1), …}
    ui.router.state: {_invokeQueue: Array(14), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(3), …}
    ui.router.util: {_invokeQueue: Array(3), _configBlocks: Array(0), _runBlocks: Array(1), info: ƒ, requires: Array(2), …}
}`

mrngoitall commented:
@petebacondarwin Is there any update on getting this PR merged?

mrngoitall commented:
@petebacondarwin Thanks!

DannyCornelisse commented:
Does anyone have a solution using the angular-cli? I'm trying to run old angularjs unit tetst using the karma config provided by angular-cli. I hope this patch will be supported soon.

AlexSwensen commented:
> Does anyone have a solution using the angular-cli? I'm trying to run old angularjs unit test using the karma config provided by angular-cli. I hope this patch will be supported soon.

Assuming your karma test files are configured properly to pickup your test directory, it should work the same as always. I would recommend opening another issue specific to your problem.

gkalpak commented on packages/upgrade/static/testing/test/mocks.ts:
```diff
@@ -0,0 +1,83 @@
+import {downgradeInjectable} from '@angular/upgrade/static';
+
+// We need to declare the Angular types as they are not exported from
+// `@angular/upgrade/...`.
```
How about using the types from [@types/angular][1]?

[1]: https://github.com/angular/angular/blob/f03b51441495d376c72ad24b506ac1ee873151d8/package.json#L44

petebacondarwin replied:
Good idea. Let me see what I can do...

gkalpak commented on packages/examples/upgrade/static/ts/full/module.spec.ts:
```diff
@@ -0,0 +1,47 @@
+
+// #docregion angular-setup
+import {TestBed} from '@angular/core/testing';
+import * as angular from '@angular/upgrade/src/common/angular1';
```
If this docregion is used in a public example, then this should not be used :grin:
How about using the actual [@types/angular][1]?

[1]: https://github.com/angular/angular/blob/f03b51441495d376c72ad24b506ac1ee873151d8/package.json#L44

gkalpak commented on packages/upgrade/static/testing/src/create_angular_testing_module.ts:
```diff
@@ -0,0 +1,98 @@
+ *
+ * In the following code snippet, `HeroesService` is an Angular service that depends upon an
+ * AngularJS
+ * service, `titleCase`.
```
Bad formatter 💠

gkalpak commented on packages/upgrade/static/testing/src/create_angular_testing_module.ts:
```diff
@@ -0,0 +1,98 @@
+ *
+ * <div class="alert is-important">
+ *
+ * The resulting configuration does not wire up AngularJS digests to Zone hooks. It is the
```
I wonder if it should (to better emulate what happens in an actual hybrid app).

gkalpak commented on packages/upgrade/static/testing/src/create_angular_testing_module.ts:
```diff
@@ -0,0 +1,98 @@
+ *
+ * @param angularJSModules a collection of the names of AngularJS modules to include in the
+ * configuration
+ * @param [strictDI] whether the AngularJS injector should have `strictDI` enabled.
```
`strictDi` for consistency with AngularJS :grin:

gkalpak commented on packages/upgrade/static/testing/src/create_angularjs_testing_module.ts:
```diff
@@ -0,0 +1,97 @@
+ * include this module in the unit test injector.
+ *
+ * In the following code snippet, we are configuring the `$injector` with two modules:
+ * The AngularJS `Ng1AppModule`, which is the AngularJS part of our hybrid application and the
```
_Ng1AppModule_ --> _ng1AppModule_

gkalpak commented on packages/upgrade/static/testing/src/create_angularjs_testing_module.ts:
```diff
@@ -0,0 +1,97 @@
+ *
+ * In the following code snippet, `heroesService` is a downgraded Angular service that we are
+ * accessing
+ * from AngularJS.
```
Bad, bad formatter.

gkalpak commented on packages/upgrade/static/testing/src/create_angularjs_testing_module.ts:
```diff
@@ -0,0 +1,97 @@
+ *
+ * @experimental
+ */
+export function createAngularJSTestingModule(AngularModules: any[]) {
```
- Shouldn't this be `Type[]`?
- Why capitalized `AngularModules`?

petebacondarwin replied:
So they probably are types but in the test bed the `TestModuleMetadata.imports` is of type `any[]` so I think I just copied that. I made them capitalized because I was thinking of them as types. But probably that is not necessary and off-putting.

mary-poppins commented:
You can preview bfe42050247959ce6d254c82cbff83eef1b12bbf at https://pr16848-bfe42050247959ce6d254c82cbff83eef1b12bbf.ngbuilds.io/.
You can preview c73b3926cbe9318a4eb56fae092d69a67e916cd0 at https://pr16848-c73b3926cbe9318a4eb56fae092d69a67e916cd0.ngbuilds.io/.
You can preview 52735ee at https://pr16848-52735ee.ngbuilds.io/.
You can preview 0743841 at https://pr16848-0743841.ngbuilds.io/.
You can preview 0260634 at https://pr16848-0260634.ngbuilds.io/.
You can preview b0a0213 at https://pr16848-b0a0213.ngbuilds.io/.
You can preview 3808f4a at https://pr16848-3808f4a.ngbuilds.io/.
You can preview 0ca84ba at https://pr16848-0ca84ba.ngbuilds.io/.
You can preview 733d0aa at https://pr16848-733d0aa.ngbuilds.io/.
You can preview 98094af at https://pr16848-98094af.ngbuilds.io/.
You can preview 37b041c at https://pr16848-37b041c.ngbuilds.io/.
You can preview 391d3b6 at https://pr16848-391d3b6.ngbuilds.io/.
You can preview 0e7ae4a at https://pr16848-0e7ae4a.ngbuilds.io/.
You can preview 78b3b07 at https://pr16848-78b3b07.ngbuilds.io/.
You can preview dd3997d at https://pr16848-dd3997d.ngbuilds.io/.
You can preview bbb982d at https://pr16848-bbb982d.ngbuilds.io/.
You can preview 12c97dd at https://pr16848-12c97dd.ngbuilds.io/.
You can preview fc00d70 at https://pr16848-fc00d70.ngbuilds.io/.
You can preview c9d70db at https://pr16848-c9d70db.ngbuilds.io/.
You can preview 3010b5d at https://pr16848-3010b5d.ngbuilds.io/.
You can preview 44a5fda at https://pr16848-44a5fda.ngbuilds.io/.
You can preview b8eedda at https://pr16848-b8eedda.ngbuilds.io/.
You can preview be9c789 at https://pr16848-be9c789.ngbuilds.io/.
You can preview 8b344a1 at https://pr16848-8b344a1.ngbuilds.io/.
You can preview 4292132 at https://pr16848-4292132.ngbuilds.io/.
You can preview 68ff3e8 at https://pr16848-68ff3e8.ngbuilds.io/.
You can preview 74a0888 at https://pr16848-74a0888.ngbuilds.io/.
You can preview f06ed88 at https://pr16848-f06ed88.ngbuilds.io/.
You can preview 5222693 at https://pr16848-5222693.ngbuilds.io/.
You can preview 98c73c4 at https://pr16848-98c73c4.ngbuilds.io/.
You can preview 74ea5a6 at https://pr16848-74ea5a6.ngbuilds.io/.
You can preview 9b3a034 at https://pr16848-9b3a034.ngbuilds.io/.
You can preview 131be9e at https://pr16848-131be9e.ngbuilds.io/.
You can preview ff417be at https://pr16848-ff417be.ngbuilds.io/.
You can preview 81a07a1 at https://pr16848-81a07a1.ngbuilds.io/.
You can preview ffbf0a9 at https://pr16848-ffbf0a9.ngbuilds.io/.
You can preview e6c00cf at https://pr16848-e6c00cf.ngbuilds.io/.

mary-poppins commented:
You can preview 2c48837 at https://pr16848-2c48837.ngbuilds.io/.

petebacondarwin commented on packages/upgrade/static/testing/BUILD.bazel:
```diff
@@ -0,0 +1,23 @@
+    #       aka. junctions are not traversed by glob.
+    srcs = glob(
+        [
+            "*./ts",
```
Ahah!

mary-poppins commented:
You can preview ba85f2e at https://pr16848-ba85f2e.ngbuilds.io/.

mary-poppins commented:
You can preview c4c98e4 at https://pr16848-c4c98e4.ngbuilds.io/.

mary-poppins commented:
You can preview b9193f6 at https://pr16848-b9193f6.ngbuilds.io/.

mary-poppins commented:
You can preview 81184b9 at https://pr16848-81184b9.ngbuilds.io/.

mary-poppins commented:
You can preview 5984a60 at https://pr16848-5984a60.ngbuilds.io/.

petebacondarwin commented:
We should also provide an "injector only version" that does not require a full TestBed to be compiled. E.g.  Something like this (thanks to @jbedard)

```
export function provideInjector(...providers: StaticProvider[]) {
    return function($provide: angular.auto.IProvideService) {
        $provide.factory("$$angularInjector", ["$injector", function($injector: angular.auto.IInjectorService) {
            return Injector.create([
                ...providers,
                {provide: "$injector", useValue: $injector}
            ]);
        }]);
        $provide.factory("$$angularLazyModuleRef", ["$$angularInjector", (injector) => ({injector, needsNgZone: true})]);
    };
}
```

theandrewlane commented:
Would be cool if this got merged in! 😬

petebacondarwin commented on packages/upgrade/static/testing/src/create_angularjs_testing_module.ts:
```diff
@@ -0,0 +1,98 @@
+            ($injector: angular.IInjectorService) => {
+              TestBed.configureTestingModule({
+                imports: angularModules,
+                providers: [{provide: $INJECTOR, useValue: $injector}]
```
Actually a small fix to //packages/compiler/src/util.ts stringify() in this PR solves this problem.

petebacondarwin committed:
fix(compiler): stringify `Object.create(null)` tokens

petebacondarwin committed:
feat(upgrade): provide unit test helpers for wiring up injectors

Adds two new helper functions that can be used when unit testing Angular services
that depend upon upgraded AngularJS services, or vice versa.
The functions return a module (AngularJS or NgModule) that is configured to wire up
the Angular and AngularJS injectors without the need to actually bootstrap a full
hybrid application.

This makes it simpler and faster to unit test services.

petebacondarwin commented on aio/tools/transforms/angular-api-package/index.js:
```diff
@@ -57,6 +57,7 @@ module.exports = new Package('angular-api', [basePackage, typeScriptPackage])
       'router/upgrade/index.ts',
       'upgrade/index.ts',
       'upgrade/static/index.ts',
+      'upgrade/testing/index.ts'
```
I moved it to `static/testing` as the dynamic version is deprecated.

petebacondarwin commented on packages/upgrade/testing/src/create_angular_test_module.ts:
````diff
@@ -0,0 +1,67 @@
+ * In the following code snippet, we are configuring the TestBed with two imports.
+ * The `AppModule` is the Angular part of our hybrid application and the `appModule` is the AngularJS part.
+ *
+ * ```
````
Now using `code-example` tags so no longer relevant.

petebacondarwin commented on packages/upgrade/testing/src/create_angularjs_test_module.ts:
````diff
@@ -0,0 +1,61 @@
+ * In the following code snippet, `shoppingCart` is an AngularJS service that depends upon an Angular service.
+ *
+ * ```
+ * it('should ...', inject(['shoppingCart', function(shoppingCart) {
````
Now using embedded examples.

petebacondarwin commented:
@pburgmer (regarding https://github.com/angular/angular/pull/16848#issuecomment-387883364). There is a bug in the `stringify()` method of the Angular compiler, which a commit in this PR now fixes.

petebacondarwin commented:
Looks like the failure is legit. I'll have to work on that.

fzxt commented:
@petebacondarwin any plans to get this in soon? do appreciate the work going on here.

petebacondarwin commented:
@fzxt - I will need help resolving the Windows CI failures...

fzxt commented on packages/upgrade/static/testing/test/create_angular_testing_module_spec.ts:
```diff
@@ -0,0 +1,47 @@
+    });
+
+    it('should create new injectors when we re-use the helper', () => {
+      TestBed.configureTestingModule({imports: [createAngularTestingModule(['app']), AppModule]});
```
Are you missing a call to `defineAppModule()` in this test?

petebacondarwin replied:
Good spot. Let me check...

petebacondarwin replied:
That was it @fzxt! Thanks.

fzxt commented:
Ping? It would be nice to get these helpers in soon. Can any googlers help out with the review?

petebacondarwin commented:
@jasonaden is working on it :-)

zuzusik commented:
any expectations on when this is going to be finished?

thanks!

petebacondarwin commented:
Apart from the conflicts it is ready to be reviewed. It is unlikely to land before 8.2 or 9 now.
I'll rebase tomorrow and chase someone (like @IgorMinar) to take a view.

gkalpak commented on packages/examples/upgrade/static/ts/full/module.ts:
```diff
@@ -13,13 +13,14 @@ import {UpgradeComponent, UpgradeModule, downgradeComponent, downgradeInjectable
 
 // #docregion ng1-text-formatter-service
-class TextFormatter {
+@Injectable()
```
I don't think this should be here 😕
This is an AngularJS service.

gkalpak commented on packages/upgrade/static/testing/BUILD.bazel:
```diff
@@ -0,0 +1,22 @@
+        "//packages/core/testing",
+        "//packages/upgrade/src/common",
+        "//packages/upgrade/static",
+        "@npm//zone.js",
```
Are all these deps needed? I only see `//packages/core` and `//packages/upgrade/src/common` being used (I think 😅).

gkalpak commented on packages/upgrade/static/testing/test/create_angularjs_testing_module_spec.ts:
```diff
@@ -0,0 +1,33 @@
+      module('app');
+      // Configure an AngularJS module that has the AngularJS and Angular injector wired up
+      module(createAngularJSTestingModule([AppModule]));
+      inject(function(shoppingCart: any) {
```
Oh, a function :grin:

petebacondarwin replied:
?

gkalpak replied:
Don't mind me :grin:
I'm just used to arrow functions these days and seeing a plain ol' function feels noteworthy 😇

gkalpak commented on packages/upgrade/static/testing/tsconfig-build.json:
```diff
@@ -0,0 +1,22 @@
+{
```
This file is now obsolete, I think.
(@alexeagle is trying to get rid of the many tsconfig files :grin:)

petebacondarwin replied:
Deleting 🤞

petebacondarwin committed:
fixup! feat(upgrade): provide unit test helpers for wiring up injectors

alxhub reviewed (approved):
+1 for compiler.

IgorMinar reviewed (approved):
Lgtm for API changes.

AndrewKushnir commented:
[VE presubmit](http://test/OCL:253626852:BASE:253626872:1560797282989:62a6accb)

AndrewKushnir commented:
Presubmit indicated that we need to perform additional changes in build files in g3 to land this PR. I've added "cleanup" label for now and will look into the build files update asap. Thank you.

petebacondarwin commented:
Thanks @AndrewKushnir

AndrewKushnir commented:
[New VE presubmit](https://test.corp.google.com/ui#id=OCL:254086143:BASE:254086863:1560984410028:df9f49c4), including g3 changes.

**Note to Caretaker:** this PR requires additional change in g3: [cl/254085499](http://cl/254085499), to update BUILD file with extra rules to build `upgrade/testing` package.

samjulien commented:
For anyone who finds this thread in the future, these helpers are available. Check out the docs here: https://angular.io/api/upgrade/static/testing
